### DISCUSSION TITLE
possibly bad python environment / cython 3.7.6 errors when running the example StatAnalysis python embedding use case on Irma

### DISCUSSION THREAD

**Discussion by jvigh** [dtcenter/METplus#2116] [Installation] [answered]
Dear METplus Help and @jprestop,

I've been trying to run the StatAnalysis python embedding use case on my RAL-maintained machine, Irma, and am getting some strange errors. I've tried hard to rid Irma of old environments related to METplus (e.g., my previous metplus_base environment had a python 3.7). I think I succeeded in this. I was able to successfully run this use case on Seneca, but when I try to run it on Irma, I get these errors. I wonder if it could be related to the MET installation (I think a RAL-maintained installation). 

Here's the relevant info for this issue.

Machine: Irma (Linux irma 4.9.0-19-amd64 #1 SMP Debian 4.9.320-3 (2022-10-14) x86_64 GNU/Linux)

Command line being run:  /usr/local/met-11.0.1/bin/stat_analysis -v 5 -lookin python /d1/personal/xxxxx/test_METplus_use_cases/read_ascii_mpr.py /d1/xxxxx/UTILITIES/METplus/METplus_Data/met_test/out/point_stat/point_stat_360000L_20070331_120000V_mpr.txt -config /d1/xxxxx/UTILITIES/METplus/METplus/parm/met_config/STATAnalysisConfig_wrapped

Error received:
ERROR  :
ERROR  : Python3_Script::Python3_Script(const char *) -> unable to open script "set_python_env"
ERROR  :
Traceback (most recent call last):
  File "/usr/local/met-11.0.1/share/met/wrappers/set_python_env.py", line 9, in <module>
    import numpy as np
  File "/d1/xxxxx/UTILITIES/miniconda/lib/python3.7/site-packages/numpy/__init__.py", line 150, in <module>
    from . import core
  File "/d1/xxxxx/UTILITIES/miniconda/lib/python3.7/site-packages/numpy/core/__init__.py", line 102, in <module>
    from . import _internal
  File "/d1/xxxxx/UTILITIES/miniconda/lib/python3.7/site-packages/numpy/core/_internal.py", line 19, in <module>
    IS_PYPY = platform.python_implementation() == 'PyPy'
  File "/d1/xxxxx/UTILITIES/miniconda/lib/python3.7/platform.py", line 1265, in python_implementation
    return _sys_version()[0]
  File "/d1/xxxxx/UTILITIES/miniconda/lib/python3.7/platform.py", line 1226, in _sys_version
    repr(sys_version))
ValueError: failed to parse CPython sys.version: '3.7.6 (default, Feb  5 2020, 11:17:15) \n[GCC 6.3.0 20170516]'

Here's the final METplus config file: 

(py_embed_base.v5) xxxxx@irma:/d1/xxxxx/UTILITIES/test_METplus_use_cases$ cat /d1/xxxxx/UTILITIES/test_METplus_use_cases/data_output/metplus_final.conf.20230412013909
[config]
PROCESS_LIST = StatAnalysis
OMP_NUM_THREADS = 1
MET_PYTHON_EXE = /d1/xxxxx/UTILITIES/miniconda/envs/py_embed_base.v5/bin/python
LOOP_BY = VALID
VALID_TIME_FMT = %Y%m%d
VALID_BEG = 20070331
VALID_END = 20070331
VALID_INCREMENT = 12H
MODEL1 = WRF
MODEL1_OBTYPE = ADPUPA
MODEL1_STAT_ANALYSIS_LOOKIN_DIR = python /d1/personal/xxxxx/test_METplus_use_cases/read_ascii_mpr.py {INPUT_BASE}/met_test/out/point_stat/point_stat_360000L_20070331_120000V_mpr.txt
MODEL1_STAT_ANALYSIS_OUT_STAT_TEMPLATE = {model?fmt=%s}_{obtype?fmt=%s}_valid{valid?fmt=%Y%m%d}_fcstvalidhour{valid_hour?fmt=%H}0000Z_out_stat.stat
STAT_ANALYSIS_OUTPUT_DIR = {OUTPUT_BASE}/met_tool_wrapper/StatAnalysis_python_embedding
STAT_ANALYSIS_CONFIG_FILE = {PARM_BASE}/met_config/STATAnalysisConfig_wrapped
STAT_ANALYSIS_JOB_NAME = aggregate_stat
STAT_ANALYSIS_JOB_ARGS = -out_line_type sl1l2 -by FCST_VAR -out_stat [out_stat_file]
MODEL_LIST = {MODEL1}
DESC_LIST =
FCST_LEAD_LIST =
OBS_LEAD_LIST =
FCST_VALID_HOUR_LIST = 12
FCST_INIT_HOUR_LIST =
OBS_VALID_HOUR_LIST =
OBS_INIT_HOUR_LIST =
FCST_VAR_LIST =
OBS_VAR_LIST =
FCST_UNITS_LIST =
OBS_UNITS_LIST =
FCST_LEVEL_LIST =
OBS_LEVEL_LIST =
VX_MASK_LIST =
INTERP_MTHD_LIST =
INTERP_PNTS_LIST =
FCST_THRESH_LIST =
OBS_THRESH_LIST =
COV_THRESH_LIST =
ALPHA_LIST =
LINE_TYPE_LIST = MPR
GROUP_LIST_ITEMS = FCST_INIT_HOUR_LIST
LOOP_LIST_ITEMS = FCST_VALID_HOUR_LIST, MODEL_LIST
STAT_ANALYSIS_SKIP_TIMES =
SKIP_TIMES =
STAT_ANALYSIS_MANDATORY = True
STAT_ANALYSIS_SKIP_IF_OUTPUT_EXISTS = False

[user_env_vars]

[runtime]
RUN_ID = 52bca9ed
CLOCK_TIME = 20230412013909
METPLUS_VERSION = 5.1.0-beta2-dev
MET_INSTALL_DIR = /usr/local/met-11.0.1
CONFIG_INPUT = /d1/xxxxx/UTILITIES/METplus/METplus/parm/metplus_config/defaults.conf,/d1/xxxxx/UTILITIES/test_METplus_use_cases/test_use_cases_irma.conf,/d1/xxxxx/UTILITIES/test_METplus_use_cases/StatAnalysis_python_embedding.conf
METPLUS_CONF = {OUTPUT_BASE}/metplus_final.conf.{LOG_TIMESTAMP}
TMP_DIR = {OUTPUT_BASE}/tmp
STAGING_DIR = {OUTPUT_BASE}/stage
FILE_LISTS_DIR = {STAGING_DIR}/file_lists.{LOG_TIMESTAMP}
CONVERT = convert
GEMPAKTOCF_JAR = GempakToCF.jar
GFDL_TRACKER_EXEC = /path/to/standalone_gfdl-vortextracker_v3.9a/trk_exec
INPUT_MUST_EXIST = True
USER_SHELL = bash
DO_NOT_RUN_EXE = False
SCRUB_STAGING_DIR = True
MET_BIN_DIR = {MET_INSTALL_DIR}/bin
INPUT_BASE = /d1/xxxxx/UTILITIES/METplus/METplus_Data
OUTPUT_BASE = /d1/xxxxx/UTILITIES/test_METplus_use_cases/data_output
LOG_METPLUS = /d1/xxxxx/UTILITIES/test_METplus_use_cases/data_output/logs/metplus.log.20230412013909
LOG_DIR = {OUTPUT_BASE}/logs
LOG_TIMESTAMP_TEMPLATE = %Y%m%d%H%M%S
LOG_TIMESTAMP_USE_DATATIME = no
LOG_MET_OUTPUT_TO_METPLUS = yes
LOG_LEVEL = INFO
LOG_LEVEL_TERMINAL = INFO
LOG_MET_VERBOSITY = 2
LOG_LINE_FORMAT = %(asctime)s.%(msecs)03dZ %(name)s (%(filename)s:%(lineno)d) %(levelname)s: %(message)s
LOG_ERR_LINE_FORMAT = {LOG_LINE_FORMAT}
LOG_DEBUG_LINE_FORMAT = {LOG_LINE_FORMAT}
LOG_INFO_LINE_FORMAT = %(asctime)s.%(msecs)03dZ %(name)s %(levelname)s: %(message)s
LOG_LINE_DATE_FORMAT = %m/%d %H:%M:%S
LOG_STAT_ANALYSIS_VERBOSITY = 5
LOG_TIMESTAMP = 20230412013909
METPLUS_BASE = /d1/xxxxx/UTILITIES/METplus/METplus
PARM_BASE = /d1/xxxxx/UTILITIES/METplus/METplus/parm

The data files being used are in the standard METplus_Data collection. Specifically: 
/d1/xxxxx/UTILITIES/METplus/METplus_Data/met_test/out/point_stat/point_stat_360000L_20070331_120000V_mpr.txt

Finally, here's my environment: 
 env
CONDA_SHLVL=1
CONDA_EXE=/d1/xxxxx/UTILITIES/miniconda/bin/conda
METPLUS_TEST_DIR=/d1/xxxxx/UTILITIES/test_METplus_use_cases
CONDA_PREFIX=/d1/xxxxx/UTILITIES/miniconda/envs/py_embed_base.v5
JAVA_HOME=/usr/local/jdk
METPLUS_BUILD_BASE=/d1/xxxxx/UTILITIES/METplus/METplus
METPLUS_DATA=/d1/projects/METplus/METplus_Data
CONDA_PYTHON_EXE=/d1/xxxxx/UTILITIES/miniconda/bin/python
_CE_CONDA=
MET_BUILD_BASE=/usr/local/met-11.0.1
CONDA_PROMPT_MODIFIER=(py_embed_base.v5)
SHELL=/bin/bash
SHLVL=1
PYTHONPATH=/usr/local/python3
PATH=/d1/xxxxx/UTILITIES/METplus/METplus/ush:/usr/local/met-11.0.1/bin:/d1/xxxxx/UTILITIES/miniconda/envs/py_embed_base.v5/bin:/d1/xxxxx/UTILITIES/miniconda/condabin:/opt/bin:/usr/local/bin:/usr/contrib/bin:/bin:/usr/bin:/usr/sbin:/usr/bin/X11:/usr/local:/home/xxxxx/mylocal/mybin:/home/xxxxx/.local/bin:/usr/local/nco/bin:/usr/local/netcdf/bin:/usr/local/ncview/bin:/home/xxxxx/mylocal/ncl_ncarg-6.6.2/bin:/usr/local/netcdf-4.3.0/gcc-4.7.2/bin:/home/xxxxx/mylocal/ecmwf_grib_api/bin:/opt/google/chrome
CONDA_DEFAULT_ENV=py_embed_base.v5
_=/usr/bin/env

Any thoughts on what to try next?

Thanks,
Jonathan

**Comment by jprestop** [comment]
Hi @jvigh. Thank you for submitting this problem as a Discussion.  I am not familiar with the machine Irma. I see that the MET installation is located at /usr/local/met-11.0.1, which leads me to believe this is the same /usr/local location as is on Seneca.  Thank you for running this same case on Seneca. The fact that it works on Seneca but not on Irma and your comment "I think a RAL-maintained installation" leads me to believe this is an environment problem, assuming the MET installation is the same. Can you please tell me which Python you were using on Seneca to test this case with?  Also it looks like you are using Python 3.7 on Irma (/d1/xxxxx/UTILITIES/miniconda/lib/python3.7/), but I am wondering based on your comment:

> I've tried hard to rid Irma of old environments related to METplus (e.g., my previous metplus_base environment had a python 3.7).

if that is not the version of Python you intend to be using.  Could you please clarify?

**Comment by jvigh** [comment]
Hi @jprestop,

I think that I'm using the exact same python on both machines: 3.8.6, as
provided by the py_embed_base.v5 conda environment:

(py_embed_base.v5) ***@***.***:/d1/xxxxx/UTILITIES/test_METplus_use_cases$
which python
/d1/xxxxx/UTILITIES/miniconda/envs/py_embed_base.v5/bin/python
(py_embed_base.v5) ***@***.***:/d1/xxxxx/UTILITIES/test_METplus_use_cases$
which python3
/d1/xxxxx/UTILITIES/miniconda/envs/py_embed_base.v5/bin/python3
(py_embed_base.v5) ***@***.***:/d1/xxxxx/UTILITIES/test_METplus_use_cases$
python
Python 3.8.6 | packaged by conda-forge | (default, Jan 25 2021, 23:21:18)
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.


(base) ***@***.***:/d1/personal/xxxxx$ conda_met pygrib.v5
(pygrib.v5) ***@***.***:/d1/personal/xxxxx$ which python
/home/met_test/.conda/envs/pygrib.v5/bin/python
(pygrib.v5) ***@***.***:/d1/personal/xxxxx$ python
Python 3.8.6 | packaged by conda-forge | (default, Jan 25 2021, 23:21:18)
[GCC 9.3.0] on linux

However, on Irma, my environment contains:
CONDA_PYTHON_EXE=/d1/xxxxx/UTILITIES/miniconda/bin/python

(py_embed_base.v5) ***@***.***:/d1/xxxxx/UTILITIES/test_METplus_use_cases$
/d1/xxxxx/UTILITIES/miniconda/bin/python
Python 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 21:52:21)
[GCC 7.3.0] :: Anaconda, Inc. on linux

Could this cause a problem? But this is still not 3.7.6, so I don't know
what is going on.

Seneca is a much newer machine than Irma (2018). Would it be guaranteed
that the MET installations are identical on both machines? Or does RAL IT
maintain different installations for different vintage machines?

Thanks,
Jonathan




Jonathan Vigh						
Project Scientist I, Joint Numerical Testbed	
Research Applications Laboratory (RAL)
National Center for Atmospheric Research (NCAR)
P.O. Box 3000                    tel: +1 (303) 497-8205
Boulder, CO 80307-3000   fax: +1 (303) 497-8171Jonathan's Staff Web
Page <http://staff.ral.ucar.edu/jvigh/> (CV, publications,
etc.)Tropical Cyclone Guidance Project
<http://hurricanes.ral.ucar.edu/> (real-time hurricane data)

Tropical Cyclone Data Project <https://verif.rap.ucar.edu/tcdata/>
(FLIGHT+, VDM+, TC-OBS datasets)

HurricaneRiskCalculator®️ <https://wxrisk.io/> (Now available --
Personalized hurricane wind hazard assessments for decision support)

During COVID-19, my working day is likely different from your working
day. Please do not feel obliged to reply to this email outside of your
normal working hours.

**Reply by jprestop** [reply]
@jvigh   

In the information you listed under "Error received:" in your initial post, you can see that Python 3.7 is being used:

> /d1/xxxxx/UTILITIES/miniconda/lib/python3.7
and 
> ValueError: failed to parse CPython sys.version: '**3.7.6** (default, Feb 5 2020, 11:17:15) \n[**GCC 6.3.0** 20170516]'

As you indicated, note the Python 3.7.6 with GCC 6.3.0.

You mentioned:
"However, on Irma, my environment contains:
CONDA_PYTHON_EXE=/d1/xxxxx/UTILITIES/miniconda/bin/python"
However, it looks like that is Python 3.7.3 with GCC 7.3.0, as you noticed as well.

I do believe these various Python environments are causing problems.  Which Python comes first in your PATH environment variable?  Is it possible to get a clean environment (with no paths to Python in your PATH) and then add the desired Python path to PATH?
e.g. export PATH=/d1/xxxxx/UTILITIES/miniconda/envs/py_embed_base.v5/bin:${PATH}

With regard to:

> Seneca is a much newer machine than Irma (2018). Would it be guaranteed
that the MET installations are identical on both machines? Or does RAL IT
maintain different installations for different vintage machines?

They usually have a separate install for each version of Debian (e.g. Debian 10 ( buster ) are Debian 9 ( stretch ) are separate installs, but I believe they're installed in the same way).

**Comment by jvigh** [comment]
@jrestop and adding @bikegeek

I'm pretty sure this must be coming from my default conda environment,
'base'.

I just ran the following to prevent the loading of that environment at
startup:
conda config --set auto_activate_base false

However, for some reason, my path is STILL being set to that other python.
CONDA_PYTHON_EXE=/d1/xxxxx/UTILITIES/miniconda/bin/python

This is python 3.7.3.

This seems to be coming from my conda.sh init script:

/d1/xxxxx/UTILITIES/miniconda/etc/profile.d

which contains the following at the top:

export CONDA_EXE='/d1/xxxxx/UTILITIES/miniconda/bin/conda'
export _CE_M=''
export _CE_CONDA=''
export CONDA_PYTHON_EXE='/d1/xxxxx/UTILITIES/miniconda/bin/python'

Is this CONDA_PYTHON_EXE causing the issue? If so, what should it be set
to? How can I get conda to give me a clean environment when I activate the
python embedding environment?

I see this:
http://progsharing.blogspot.com/2019/07/deactivate-condas-base-environment-on.html
Trick #3 recommends commenting out part of the conda init in a user's
.bashrc. Is this really necessary?

Thanks,
Jonathan




Jonathan Vigh						
Project Scientist I, Joint Numerical Testbed	
Research Applications Laboratory (RAL)
National Center for Atmospheric Research (NCAR)
P.O. Box 3000                    tel: +1 (303) 497-8205
Boulder, CO 80307-3000   fax: +1 (303) 497-8171Jonathan's Staff Web
Page <http://staff.ral.ucar.edu/jvigh/> (CV, publications,
etc.)Tropical Cyclone Guidance Project
<http://hurricanes.ral.ucar.edu/> (real-time hurricane data)

Tropical Cyclone Data Project <https://verif.rap.ucar.edu/tcdata/>
(FLIGHT+, VDM+, TC-OBS datasets)

HurricaneRiskCalculator®️ <https://wxrisk.io/> (Now available --
Personalized hurricane wind hazard assessments for decision support)

During COVID-19, my working day is likely different from your working
day. Please do not feel obliged to reply to this email outside of your
normal working hours.

**Reply by jprestop** [reply]
Thanks @jvigh. I don't have a great knowledge of conda environments so hopefully @bikegeek can help.  Let us know if a screen sharing session could help.  Perhaps together, Minna and you and I can get it working.

**Comment by jvigh** [comment]
@jprestop and @bikegeek

Sure, I'm happy to do a screen share. Please let me know when is a good
time for you.

Jonathan Vigh						
Project Scientist I, Joint Numerical Testbed	
Research Applications Laboratory (RAL)
National Center for Atmospheric Research (NCAR)
P.O. Box 3000                    tel: +1 (303) 497-8205
Boulder, CO 80307-3000   fax: +1 (303) 497-8171Jonathan's Staff Web
Page <http://staff.ral.ucar.edu/jvigh/> (CV, publications,
etc.)Tropical Cyclone Guidance Project
<http://hurricanes.ral.ucar.edu/> (real-time hurricane data)

Tropical Cyclone Data Project <https://verif.rap.ucar.edu/tcdata/>
(FLIGHT+, VDM+, TC-OBS datasets)

HurricaneRiskCalculator®️ <https://wxrisk.io/> (Now available --
Personalized hurricane wind hazard assessments for decision support)

During COVID-19, my working day is likely different from your working
day. Please do not feel obliged to reply to this email outside of your
normal working hours.

**Reply by jprestop** [reply]
Let's see if @bikegeek has any suggestions for you to try out first and go from there if that's ok with you.

**Comment by bikegeek** [comment]
let me know when you are available.

---------------
Minna Win
Pronouns: she/her
National Center for Atmospheric Research
Research Applications Lab
Developmental Testbed Center
Phone: 303-497-8423
Fax:   303-497-8401
---------------

**Comment by jvigh** [comment]
@bikegeek I'm available now. Does that work for you?

Jonathan Vigh						
Project Scientist I, Joint Numerical Testbed	
Research Applications Laboratory (RAL)
National Center for Atmospheric Research (NCAR)
P.O. Box 3000                    tel: +1 (303) 497-8205
Boulder, CO 80307-3000   fax: +1 (303) 497-8171Jonathan's Staff Web
Page <http://staff.ral.ucar.edu/jvigh/> (CV, publications,
etc.)Tropical Cyclone Guidance Project
<http://hurricanes.ral.ucar.edu/> (real-time hurricane data)

Tropical Cyclone Data Project <https://verif.rap.ucar.edu/tcdata/>
(FLIGHT+, VDM+, TC-OBS datasets)

HurricaneRiskCalculator®️ <https://wxrisk.io/> (Now available --
Personalized hurricane wind hazard assessments for decision support)

During COVID-19, my working day is likely different from your working
day. Please do not feel obliged to reply to this email outside of your
normal working hours.

**Comment by bikegeek** [comment]
@jvigh, I  created a google meeting and invited you...
---------------
Minna Win
Pronouns: she/her
National Center for Atmospheric Research
Research Applications Lab
Developmental Testbed Center
Phone: 303-497-8423
Fax:   303-497-8401
---------------

**Reply by jprestop** [reply]
@jvigh and @bikegeek, I'm sorry,  unfortunately I am not available now. I'm typing on my phone while picking up my daughter from school (parked).

**Comment by bikegeek** [comment]
@jpresto, @jvigh I am available tomorrow, pretty much any time.  Let me
know what times work for you and I'll create a Google meeting.

---------------
Minna Win
Pronouns: she/her
National Center for Atmospheric Research
Research Applications Lab
Developmental Testbed Center
Phone: 303-497-8423
Fax:   303-497-8401
---------------

**Comment by jvigh** [comment]
Sorry, I got distracted by a wildfire. I can meet now if you're still
available.


Jonathan Vigh						
Project Scientist I, Joint Numerical Testbed	
Research Applications Laboratory (RAL)
National Center for Atmospheric Research (NCAR)
P.O. Box 3000                    tel: +1 (303) 497-8205
Boulder, CO 80307-3000   fax: +1 (303) 497-8171Jonathan's Staff Web
Page <http://staff.ral.ucar.edu/jvigh/> (CV, publications,
etc.)Tropical Cyclone Guidance Project
<http://hurricanes.ral.ucar.edu/> (real-time hurricane data)

Tropical Cyclone Data Project <https://verif.rap.ucar.edu/tcdata/>
(FLIGHT+, VDM+, TC-OBS datasets)

HurricaneRiskCalculator®️ <https://wxrisk.io/> (Now available --
Personalized hurricane wind hazard assessments for decision support)

During COVID-19, my working day is likely different from your working
day. Please do not feel obliged to reply to this email outside of your
normal working hours.

**Comment by jvigh** [comment, marked as answer]
Minna and I dug into this to try to rule out any conda environment issues.
After trying many things, we ultimately found the cause:

I had the following in my .bash_profile:
export PYTHONPATH="/usr/local/python3"
This pointed to an old python 3.5.

Possibly, this older python didn't allow MET to call the set_python_envs
script, which is one of the initial errors we were getting. We don't know
exactly how it was using python/cython 3.7.6.

Additionally, we updated to miniconda3, rather than the old junky miniconda
that I was using.

Issue solved!












Jonathan Vigh						
Project Scientist I, Joint Numerical Testbed	
Research Applications Laboratory (RAL)
National Center for Atmospheric Research (NCAR)
P.O. Box 3000                    tel: +1 (303) 497-8205
Boulder, CO 80307-3000   fax: +1 (303) 497-8171Jonathan's Staff Web
Page <http://staff.ral.ucar.edu/jvigh/> (CV, publications,
etc.)Tropical Cyclone Guidance Project
<http://hurricanes.ral.ucar.edu/> (real-time hurricane data)

Tropical Cyclone Data Project <https://verif.rap.ucar.edu/tcdata/>
(FLIGHT+, VDM+, TC-OBS datasets)

HurricaneRiskCalculator®️ <https://wxrisk.io/> (Now available --
Personalized hurricane wind hazard assessments for decision support)

During COVID-19, my working day is likely different from your working
day. Please do not feel obliged to reply to this email outside of your
normal working hours.

**Reply by jprestop** [reply]
Thank you @bikegeek and @jvigh for finding the problem and solution and for updating here.  I will go ahead and mark this as answered and close it.
